vm/src/bench/kernel.rs

```diff
@@ -0,0 +1,421 @@
+    ExecutionError, GasOps, IpldBlockOps, LimiterOps, MessageOps, NetworkOps, RandomnessOps,
+    SelfOps, SendResult,
+};
+use fvm::machine::limiter::MemoryLimiter;
+
+use fvm::{DefaultKernel, Kernel};
+use fvm_shared::address::{Address, SECP_PUB_LEN};
+use fvm_shared::clock::ChainEpoch;
+use fvm_shared::consensus::ConsensusFault;
+use fvm_shared::crypto::signature::{SignatureType, SECP_SIG_LEN, SECP_SIG_MESSAGE_HASH_SIZE};
+use fvm_shared::econ::TokenAmount;
+
+use fvm_shared::piece::PieceInfo;
+use fvm_shared::randomness::RANDOMNESS_LENGTH;
+use fvm_shared::sector::{
+    AggregateSealVerifyProofAndInfos, RegisteredSealProof, ReplicaUpdateInfo, SealVerifyInfo,
+    WindowPoStVerifyInfo,
+};
+use fvm_shared::sys::out::network::NetworkContext;
+use fvm_shared::sys::out::vm::MessageContext;
+use fvm_shared::sys::SendFlags;
+use fvm_shared::{ActorID, MethodNum, TOTAL_FILECOIN};
+
+use multihash::MultihashGeneric;
+
+pub const TEST_VM_RAND_ARRAY: [u8; 32] = [
+    1u8, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25,
+    26, 27, 28, 29, 30, 31, 32,
+];
+
+// TODO: extend test configuration to allow dynamic toggling for certain features to be intercepted
+// and the behaviour once intercepted
+// https://github.com/anorth/fvm-workbench/issues/9
+#[derive(Debug, Clone)]
+pub struct TestConfiguration {
+    circ_supply: TokenAmount,
+    price_list: PriceList,
+}
+
+pub type Result<T> = std::result::Result<T, ExecutionError>;
+
+pub struct BenchKernel<C: CallManager> {
+    inner_kernel: DefaultKernel<C>,
+    test_config: TestConfiguration,
+}
+
+impl<C> Kernel for BenchKernel<C>
+where
+    C: CallManager,
+{
+    type CallManager = C;
+
+    fn into_inner(self) -> (Self::CallManager, BlockRegistry)
+    where
+        Self: Sized,
+    {
+        self.inner_kernel.into_inner()
+    }
+
+    fn new(
+        mgr: Self::CallManager,
+        blocks: BlockRegistry,
+        caller: ActorID,
+        actor_id: ActorID,
+        method: MethodNum,
+        value_received: TokenAmount,
+        read_only: bool,
+    ) -> Self
+    where
+        Self: Sized,
+    {
+        let default_kernel =
+            DefaultKernel::new(mgr, blocks, caller, actor_id, method, value_received, read_only);
+        let price_list = default_kernel.price_list().clone();
+        BenchKernel {
+            inner_kernel: default_kernel,
+            test_config: TestConfiguration { circ_supply: TOTAL_FILECOIN.clone(), price_list },
+        }
+    }
+
+    fn machine(&self) -> &<Self::CallManager as fvm::call_manager::CallManager>::Machine {
+        self.inner_kernel.machine()
+    }
+
+    fn send<K: Kernel<CallManager = Self::CallManager>>(
+        &mut self,
+        recipient: &Address,
+        method: u64,
+        params: BlockId,
+        value: &TokenAmount,
+        gas_limit: Option<Gas>,
+        flags: SendFlags,
+    ) -> fvm::kernel::Result<SendResult> {
```

Comment:
Please leave a long descriptive comment that `K` is ignored here, so this BenchKernel can't itself be wrapped. If we could pass K through to inner_kernel::send, we would, but we haven't figured out how to express that.